modules/core/src/smithy4s/http/internals/SchemaVisitorMetadataReader.scala

```diff
@@ -0,0 +1,304 @@
+      case Primitive.PDouble => MetaDecode.from("Double")(_.toDoubleOption)
+      case Primitive.PBigInt => MetaDecode.fromUnsafe("BigInt")(BigInt(_))
+      case Primitive.PBigDecimal =>
+        MetaDecode.fromUnsafe("BigDecimal")(BigDecimal(_))
+      case Primitive.PBoolean => MetaDecode.from("Boolean")(_.toBooleanOption)
+      case Primitive.PString  => MetaDecode.fromUnsafe("String")(identity)
+      case Primitive.PUUID =>
+        MetaDecode.fromUnsafe[ju.UUID]("UUID")(ju.UUID.fromString)
+      case Primitive.PByte => EmptyMetaDecode
+      case Primitive.PBlob =>
+        MetaDecode.fromUnsafe("Bytes")(string =>
+          ByteArray(ju.Base64.getDecoder().decode(string))
+        )
+      case Primitive.PDocument => EmptyMetaDecode
+      case Primitive.PTimestamp =>
+        (
+          hints.get(HttpBinding).map(_.tpe),
+          hints.get(smithy.api.TimestampFormat)
+        ) match {
+          case (_, Some(format)) =>
+            MetaDecode.from(Timestamp.showFormat(format))(str =>
+              Timestamp.parse(str, format)
+            )
+          case (Some(HttpBinding.Type.QueryType), None) |
+              (Some(HttpBinding.Type.PathType), None) =>
+            val formatString =
+              Timestamp.showFormat(smithy.api.TimestampFormat.DATE_TIME)
+            // See https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html?highlight=httpquery#httpquery-trait
+            MetaDecode.from(formatString)(str =>
+              Timestamp.parse(str, smithy.api.TimestampFormat.DATE_TIME)
+            )
+          case (Some(HttpBinding.Type.HeaderType), None) =>
+            val formatString =
+              Timestamp.showFormat(smithy.api.TimestampFormat.HTTP_DATE)
+            // See https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html?highlight=httpquery#httpheader-trait
+            MetaDecode.from(formatString)(str =>
+              Timestamp.parse(str, smithy.api.TimestampFormat.HTTP_DATE)
+            )
+          case (None, None) =>
+            EmptyMetaDecode
+        }
+      case Primitive.PUnit =>
+        MetaDecode.StructureMetaDecode(
+          _ => Right(MMap.empty[String, Any]),
+          Some(_ => Right(()))
+        )
+    }
+  }
+
+  override def list[A](
+      shapeId: ShapeId,
+      hints: Hints,
+      member: Schema[A]
+  ): MetaDecode[List[A]] = {
+    self(member) match {
+      case MetaDecode.StringValueMetaDecode(f) => {
+        MetaDecode.StringListMetaDecode[List[A]] { iterator =>
+          val buffer = ListBuffer.empty[A]
+          iterator.foreach(string => buffer += f(string))
+          buffer.toList
+        }
+      }
+      case _ => EmptyMetaDecode
+    }
+  }
+
+  override def set[A](
+      shapeId: ShapeId,
+      hints: Hints,
+      member: Schema[A]
+  ): MetaDecode[Set[A]] =
+    self(member) match {
+      case MetaDecode.StringValueMetaDecode(f) =>
+        MetaDecode.StringListMetaDecode[Set[A]] { iterator =>
```

Comment:
the comment is related to the original code I guess (because I assume you ported that from the Schematic version), but do we really have to use a mutable `ListBuffer`?

just below, in the `map` implementation, we use `map` and `toMap`. Can't we just do `iterator.map(f).toList` and `iterator.map(f).toSet` in both case?

Reply:
fair point